freeze_room/__init__.py

```diff
@@ -313,3 +343,176 @@ def _is_local_user(self, user_id: str) -> bool:
+
+    # Do some validation checks on the power level state event
+    if (
+        not isinstance(power_level_content, dict)
+        or "users" not in power_level_content
+        or not isinstance(power_level_content["users"], dict)
+    ):
+        # We can't use this power level event to determine whether the room should be
+        # frozen. Bail out.
+        return None
+
+    return power_level_content
+
+
+def _get_users_with_highest_nondefault_pl(
+    users_dict: dict,
+    users_default_pl: int,
+    state_events: StateMap[EventBase],
+    ignore_user: str,
+) -> Tuple[str]:
+    """Looks at the provided bits of power levels event content to figure out what the
+    maximum user-specific non-default power level is with users still in the room (or
+    invited to it) and which users have it.
+
+    Args:
+        users_dict: The "users" dictionary from the power levels event content.
+        users_default_pl: The default power level for users who don't appear in the users
+            dictionary.
+        state_events: The current state of the room, from which we can check the room's
+            member list.
+        ignore_user: A user to ignore, i.e. to consider they've left the room even if the
+            room's state says otherwise.
+
+    Returns:
+        A tuple of users with the highest non-default power level, or an empty set if no
+        such users exist in the room.
+    """
+    # Make a copy of the users dict so we don't modify the actual event content.
+    users_dict_copy = users_dict.copy()
+
+    if ignore_user in users_dict_copy:
+        del users_dict_copy[ignore_user]
+
+    while True:
+        # If there's no more user to evaluate, return an empty tuple.
+        if not users_dict_copy:
+            return tuple()
+
+        # Get the max power level in the dict.
+        max_pl = max(users_dict_copy.values())
+
+        # Bail out if the max power level is the default one (or is lower).
+        if max_pl <= users_default_pl:
+            return tuple()
+
+        # Figure out which users will need promoting: every user with the max power level
+        # that's still in the room (or have a pending invite to it).
+        users_to_promote = tuple(
+            user_id
+            for user_id, pl in users_dict_copy.items()
+            if pl == max_pl
+            and (
+                _get_membership(user_id, state_events)
+                in [Membership.JOIN, Membership.INVITE]
+            )
+        )
+
+        # If we've got users in the room to promote, break out and return.
+        if users_to_promote:
+            return users_to_promote
+
+        # Otherwise, remove the users we've considered and start again.
+        # We do it in two loops because we can't delete from users_dict_copy while
+        # iterating over it.
+        users_to_delete = []
+        for user_id, pl in users_dict_copy.items():
+            if pl == max_pl:
+                users_to_delete.append(user_id)
```

Comment:
This is theoretically more efficient, and personally I find it easier to read; ymmv though.

```suggestion
        users_to_delete = tuple(
            user_id 
            for user_id, pl in users_dict_copy.items() 
            if pl == max_pl
        )
```

Another thought: do this *before* checking the membership states (and call it  `max_pl_users` or somesuch instead of `users_to_delete`). Then iterate over that instead of `users_dict_copy` to build `users_to_promote`.